tests/test_feature_price_emission.py

```diff
@@ -0,0 +1,203 @@
+    years = [2020, 2025, 2030, 2040]
+
+    model_setup(scen, years)
+    for y in years:
+        scen.add_cat('year', y, y)
+        scen.add_par('bound_emission',
+                     ['World', 'ghg', 'all', y], 0, 'tCO2')
+    scen.commit('initialize test scenario')
+    scen.solve()
+
+    # with an emissions constraint, the technology with costs satisfies demand
+    assert scen.var('OBJ')['lvl'] > 0
+    # under per-year emissions constraints, the emission price must be equal to
+    # the marginal relaxation, ie. the difference in costs between technologies
+    npt.assert_allclose(scen.var('PRICE_EMISSION')['lvl'].values, [1] * 4)
+
+
+def test_custom_type_variable_periodlength(test_mp):
+    scen = Scenario(test_mp, MODEL, 'cum_equidistant', version='new')
+    years = [2020, 2025, 2030, 2040, 2050]
+    custom = [2025, 2030, 2040]
+
+    model_setup(scen, years)
+    scen.add_cat('year', 'custom', custom)
+    scen.add_par('bound_emission',
+                 ['World', 'ghg', 'all', 'custom'], 0, 'tCO2')
+
+    scen.commit('initialize test scenario')
+    scen.solve()
+
+    # with an emissions constraint, the technology with costs satisfies demand
+    assert scen.var('OBJ')['lvl'] > 0
+    # under a cumulative constraint, the price must increase with the discount
+    # rate starting from the marginal relaxation in the first year
+    obs = scen.var('PRICE_EMISSION')['lvl'].values
+    npt.assert_allclose(obs, [1.05**(y - custom[0]) for y in custom])
+
+
+def test_price_duality(test_mp):
+    years = [2020, 2025, 2030, 2040, 2050]
+    for c in [0.25, 0.5, 0.75]:
+        # set up a scenario for cumulative constraints
+        scen = Scenario(test_mp, MODEL, 'cum_many_tecs', version='new')
+        model_setup(scen, years, simple_tecs=False)
+        scen.add_cat('year', 'cumulative', years)
+        scen.add_par('bound_emission',
+                     ['World', 'ghg', 'all', 'cumulative'], 0.5, 'tCO2')
+        scen.commit('initialize test scenario')
+        scen.solve()
+
+        # set up a new scenario with emissions taxes
+        tax_scen = Scenario(test_mp, MODEL, 'tax_many_tecs', version='new')
+        model_setup(tax_scen, years, simple_tecs=False)
+        for y in years:
+            tax_scen.add_cat('year', y, y)
+
+        # use emission prices from cumulative-constraint scenario as taxes
+        taxes = scen.var('PRICE_EMISSION')\
+            .rename(columns={'year': 'type_year', 'lvl': 'value'})
+        taxes['unit'] = 'USD/tCO2'
+        tax_scen.add_par('tax_emission', taxes)
+        tax_scen.commit('initialize test scenario for taxes')
+        tax_scen.solve()
+
+        # check that emissions are close between cumulative and tax scenario
+        filters = {'node': 'World'}
+        emiss = scen.var('EMISS', filters).set_index('year').lvl
+        emiss_tax = tax_scen.var('EMISS', filters).set_index('year').lvl
+        npt.assert_allclose(emiss, emiss_tax, rtol=0.20)
```

Comment:
isn't a 20% tolerance really high?

Reply:
Agreed in person to add a comment here explaining why 20% is necessary

Reply:
Because of the linear model, setting the dual from the emission-bound-scenario as a price will result in different emissions trajectories. This was the closest test that I could come up with in a reasonable amount of time.

Having a larger number of technologies would likely reduce the gap, but run afoul of the GAMS test license model size limitation.